src/lib/core/evio.c

```diff
@@ -367,59 +392,178 @@ evio_service_bind(struct evio_service *service, const char *uri)
 
 	if (close(service_fd) < 0)
 		say_error("Failed to close socket: %s", strerror(errno));
 
-	if (service->addr.sa_family != AF_UNIX)
+	if (entry->addr.sa_family != AF_UNIX)
 		return;
 
-	if (unlink(((struct sockaddr_un *)&service->addr)->sun_path) < 0) {
+	if (unlink(((struct sockaddr_un *)&entry->addr)->sun_path) < 0) {
 		say_error("Failed to unlink unix "
 			  "socket path: %s", strerror(errno));
 	}
 }
 
+static void
+evio_service_entry_attach(struct evio_service_entry *dst,
+			 const struct evio_service_entry *src)
+{
+	strcpy(dst->host, src->host);
+	strcpy(dst->serv, src->serv);
+	dst->addrstorage = src->addrstorage;
+	dst->addr_len = src->addr_len;
+	ev_io_set(&dst->ev, src->ev.fd, EV_READ);
+}
+
+static inline int
+evio_service_reuse_addr(const char **uris, int size)
+{
+	for (int i = 0; i < size; i++)
+		if (evio_service_entry_reuse_addr(uris[i]) != 0)
+			return -1;
+	return 0;
+}
+
+char **
+evio_service_bound_address(const struct evio_service *service, int *size)
+{
+	*size = service->entry_count;
+	if (service->entry_count == 0)
+		return NULL;
+	char **bound_address_array = (char **)
+		xcalloc(service->entry_count,
+			sizeof(char *) + SERVICE_NAME_MAXLEN);
+	char *dstbuf = (char *)bound_address_array +
+		service->entry_count * sizeof(char *);
+	for (int i = 0; i < service->entry_count; i++) {
+		bound_address_array[i] =
+			dstbuf + i * SERVICE_NAME_MAXLEN;
+		evio_service_entry_bound_address(&service->entries[i],
+						 bound_address_array[i]);
+	}
+	return bound_address_array;
+}
+
+void
+evio_service_bound_address_free(char **bound_address_array)
+{
+	free(bound_address_array);
+}
+
+void
+evio_service_init(struct ev_loop *loop, struct evio_service *service, int size,
+		  const char *name, evio_accept_f on_accept,
+		  void *on_accept_param)
+{
+	service->entries = (size != 0 ?
+		 xcalloc(size, sizeof(struct evio_service_entry)) : NULL);
+	service->entry_count = size;
+	snprintf(service->name, sizeof(service->name), "%s", name);
+	service->loop = loop;
+	service->on_accept = on_accept;
+	service->on_accept_param = on_accept_param;
+	for (int i = 0; i < service->entry_count; i++)
+		evio_service_entry_init(&service->entries[i], service);
+}
+
+void
+evio_service_attach(struct evio_service *dst, const struct evio_service *src)
+{
+	for (int i = 0; i < src->entry_count; i++)
+		evio_service_entry_attach(&dst->entries[i], &src->entries[i]);
+	dst->entry_count = src->entry_count;
+}
+
 void
 evio_service_detach(struct evio_service *service)
 {
-	if (ev_is_active(&service->ev)) {
-		ev_io_stop(service->loop, &service->ev);
-		service->addr_len = 0;
+	for (int i = 0; i < service->entry_count; i++)
+		evio_service_entry_detach(&service->entries[i]);
+	free(service->entries);
+	service->entry_count = 0;
+	service->entries = NULL;
+}
+
+bool
+evio_service_is_active(struct evio_service *service)
+{
+	for (int i = 0; i < service->entry_count; i++)
+		if (evio_service_entry_is_active(&service->entries[i]))
+			return true;
```

Comment:
You would need `{}` for the loop body when it has > 1 line. Also `service` can be made `const` here. Together with entry argument in `evio_service_entry_is_active`.